arm: tfm: samples: psa level 1 certified sample

microbuilder opened the pull request:
**WORK IN PROGRESS, SHOULD NOT BE MERGED!**

This PR adds some boilerplate code for a PSA Certified Level 1
sample application that can be used during the PSA certification
process.

It implements the following features:

- A config struct is defined (`cfg_data`) that contains app and user
  configuration data that will be persisted to secure storage via the
  appropriate PSA FF API calls.
- A `psa` shell command is added that can be extended with required
  features, such as the ability to delete the above config record and
  reset the device to factory defaults.
- Basic logging support to handle common TF-M error codes (`applog.c`),
  based on the Zephyr logging subsystem.
- Basic encryption/decryption support showing how the TF-M crypto
  service can be used to encrypt and decrypt data transfers.

This sample depends on #16450.

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>

Commit:
arm: mps_an521_nonsecure: Add TFM BL2 support

PSA level 1 requires secure boot. TF-M BL2 is the official
secure boot loader. It needs a BL2_HEADER_SIZE offset.

Align nonsecure address with TF-M's NS slot while TF-M BL2 enabled.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

microbuilder committed:
arm: tfm: samples: psa level 1 certified sample

This commit adds some boilerplate code for a PSA Certified Level 1
sample application that can be used during the PSA certification
process.

It implements the following features:

- A config struct is defined (`cfg_data`) that contains app and user
  configuration data that will be persisted to secure storage via the
  appropriate PSA FF API calls.
- A `psa` shell command is added that can be extended with required
  features, such as the ability to delete the above config record and
  reset the device to factory defaults.
- Basic logging support to handle common TF-M error codes (`applog.c`),
  based on the Zephyr logging subsystem.
- Basic encryption/decryption support showing how the TF-M crypto
  service can be used to encrypt and decrypt data transfers.

This sample is a **WORK IN PROGRESS** and **SHOULD NOT BE MERGED**.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: tfm: sample: checkpatch/doc fixes

This commit fixed several checkpath and doc issues with the PR.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: tfm: sample: checkpatch/doc fixes

This commit fixed several checkpath and doc issues with the PR.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+Overview
+********
+This TF-M integration example can be used with an Arm v8-M supported board. It
+implements the RTOS vendor requirements for a PSA Certified Level 1 product.
```
add the link you defined below:
```
implements the RTOS vendor requirements for a `PSA Certified Level 1`_ product.
```

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+stored in the secure bootloader, so that they will be accepted during the
+image validation process at startup.
+
+All TF-M dependencies will be automatically cloned in the /ext folder, and
```
change to:
```
in the ``/ext`` folder
```

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+This project outputs startup status and info to the console. It can be built and
+executed on an MPS2+ configured for AN521 (dual-core ARM Cortex M33).
+
+At present, this sample is only compatible with Linux and OS X based systems,
```
Change to:
```
This sample will only build on a Linux or macOS development system (not Windows),
```

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+At present, this sample is only compatible with Linux and OS X based systems,
+and has been tested on the following setups:
+
+- OS X 10.14.5 natively with gcc-arm-none-eabi-7-2018-q2-update
```
change ``OS X`` to ``macOS Mojave``

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+On MPS2+ AN521:
+===============
+
+1. Build Zephyr with a non-secure configuration (``-DBOARD=mps2_an521_nonsecure``). This will also build TF-M in the ``ext/tfm`` folder if necessary.
```
Keep line length < 80 when possible

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+
+1. Build Zephyr with a non-secure configuration (``-DBOARD=mps2_an521_nonsecure``). This will also build TF-M in the ``ext/tfm`` folder if necessary.
+
+Using ``west``
```
All these lines (46-61) should be indented an additional three spaces to be part of step 1.

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+
+2. Merge the TF-M secure and Zephyr non-secure images, and sign the final image
+
+.. code-block:: bash
```
same here.  Indent the code block and contents (line 65-67) by an additional three spaces.

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+   ./merge.sh
+
+3. Copy application binary files (mcuboot.bin and tfm_sign.bin) to ``<MPS2 device name>/SOFTWARE/``.
+4. Open ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``.
```
Let's merge some of these steps and fix some formatting issues:
```
4. Edit (e.g., with vim) the ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``
   file, and update it as shown below:

   .. code-block:: bash

      TITLE: Versatile Express Images Configuration File

      [IMAGES]
      TOTALIMAGES: 2 ;Number of Images (Max: 32)

      IMAGE0ADDRESS: 0x10000000
      IMAGE0FILE: \SOFTWARE\mcuboot.bin  ; BL2 bootloader

      IMAGE1ADDRESS: 0x10080000
      IMAGE1FILE: \SOFTWARE\tfm_sign.bin ; TF-M with application binary blob

5. Save the file, exit the editor, and reset the MPS2+ board.
```

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+firmware image is validated by the bootloader at startup using the public key,
+which is built into the secure bootloader.
+
+By default, tfm/bl2/ext/mcuboot/root-rsa-3072.pem is used to sign images.
```
It renders better if you enclose filenames with double back-ticks:
```
By default, ``tfm/bl2/ext/mcuboot/root-rsa-3072.pem`` is used to sign images.
The ``merge.sh`` script signs the TF-M + Zephyr binary using the ``.pem`` private key,
by calling ``imgtool.py`` to perform the actual signing operation.
```

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+merge.sh signs the TF-M + Zephyr binary using the .pem private key,
+calling imgtool.py` to perform the actual signing operation.
+
+In order to satisfy PSA Level 1 certification requirements, **You MUST replace
```
Change ``In order to`` to simply ``To``

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,145 @@
+   ...
+
+
+.. _TF-M build instruction:
```
You've defined these two external links, but don't use them anywhere in the document so they won't show up in our generated HTML.  Note what I did on line 9 above to add a link in the documentation to the document you identified.  We need to create similar references in the text to these two documents.

Commit:
test cases: crypto: Call TFM crypto service in NS

Add TFM test suite for crypto in nonsecure as test cases from Zephyr.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

karl-zh commented on boards/arm/mps2_an521/mps2_an521_nonsecure.dts:
```diff
@@ -45,8 +45,8 @@
 
-	flash0: flash@100000 {
-		reg = <0x100000 0xDF00000>;
+	flash0: flash@100400 {
```
This commit is from #16450 at [here](https://github.com/zephyrproject-rtos/zephyr/pull/16450/commits/f9d10e31e2436c2dc604f2a3fe43d8dd2b67e961) to resolve secure fault while jump into Zephyr.

karl-zh reviewed (commented):
The Crypto test cases are leveraging the same non secure test cases from TFM. These test cases are designed for TFM crypto service. @microbuilder Is it fine to take advantage these integration test cases here?

microbuilder committed:
arm: tfm: samples: PSA level 1 sample app improvements

This commit adds basic support for working with the initial
attestation service in TF-M, as well as a rework of thee config and
logging system.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm: checkpatch improvements

This commit addresses some spacing issues found via checkpatch.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm: additional checkpatch cleanup

This commit resolves further checkpatch and stylistic issues with
this PR.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: tfm: samples: IAT public key fix

This commit adds a sample public key to demonstrate how to handle
IAT signature verification, and some minor documentation updates on
how to use the check_iat tool to this end.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm: Fix links in documentation

This commit fixes the links in attestation.md to properly use
deferred link URIs.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm: additional IAT documentation

This commit improves the IAT documentation with details on the packet
contents via the decompile_token command.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm: Note on further attestation schemes

This commit adds some notes on alternative uses of the IAT, such as
attesting to sensor data in an asynchronous or unidirectional network
setup.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm add SHA256 hash sample

This commit adds an example of how to use the PSA Crypto API to
calculate the SHA-256 hash of a payload in chunks.

It also includes a handful of minor documentation cleanups.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder committed:
arm: samples: tfm: Additional crypto examples

This commit adds a dedicated file for psa_crypto examples.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder commented:
Superseded by #19985, which takes into account the changes and feedback made in this PR.

This PR is being closed in favour of the newer version which integrates a shared TF-M build system, and will simplify the overall TF-M sample app review process.
